text/0002-observables.md

```diff
@@ -0,0 +1,461 @@
+
+```js
+<script>
+  import { user } from './stores.js';
+</script>
+
+<h1>Hello {$user.firstname}!</h1>
+```
+
+This would compile to something like the following:
+
+```js
+import { onDestroy } from 'svelte';
+import { user } from './stores.js';
+
+function init($$self, $$make_dirty) {
+  let $user;
+  onDestroy(user.subscribe(value => {
+    $user = value;
+    $$makeDirty('$user');
+  }));
+
+  $$self.get = () => ({ $user });
+}
+```
+
+
+### Store API
+
+A store *must* have a `subscribe` method, and it may also have additional methods like `set` and `update` if it isn't read-only:
+
+```js
+const number = writable(1);
+
+const unsubscribe = number.subscribe(value => {
+  console.log(`value is ${value}`); // logs 1 immediately
+});
+
+number.set(2); // logs 2
+
+const incr = n => n + 1;
+number.update(incr); // logs 3
+
+unsubscribe();
+number.set(4); // does nothing — unsubscribed
+```
+
+An example implementation of this API:
+
+```js
+function writable(value) {
+  const subscribers = [];
+
+  function set(newValue) {
+    if (newValue === value) return;
+    value = newValue;
+    subscribers.forEach(fn => fn(value));
+  }
+
+  function update(fn) {
+    set(fn(value));
+  }
+
+  function subscribe(fn) {
+    subscribers.push(fn);
+    fn(value);
+
+    return function() {
+      const index = subscribers.indexOf(fn);
+      if (index !== -1) subscribers.splice(index, 1);
+    }
+  }
+
+  return { set, update, subscribe };
+}
+```
+
+
+### Read-only stores
+
+Some stores are read-only, created with `readable`:
+
+```js
+const unsubscribe = mousePosition.subscribe(pos => {
+  if (pos) console.log(pos.x, pos.y);
+});
+
+mousePosition.set({ x: 100, y: 100 }); // Error: mousePosition.set is not a function
+```
+
+An example implementation:
+
+```js
+function readable(start, value) {
+  const subscribers = [];
+  let stop;
+
+  function set(newValue) {
+    if (newValue === value) return;
+    value = newValue;
+    subscribers.forEach(fn => fn(value));
+  }
+
+  return {
+    subscribe(fn) {
+      if (subscribers.length === 0) {
+        stop = start(set);
+      }
+
+      subscribers.push(fn);
+      fn(value);
+
+      return function() {
+        const index = subscribers.indexOf(fn);
+        if (index !== -1) subscribers.splice(index, 1);
+
+        if (subscribers.length === 0) {
+          stop();
+          stop = null;
+        }
+      };
+    }
+  }
+}
+
+const mousePosition = readable(function start(set) {
+  function handler(event) {
+    set({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  document.addEventListener('mousemove', handler);
+  return function stop() {
+    document.removeEventListener('mousemove', handler);
+  }
+});
+```
+
+### Derived stores
+
+A store can be derived from other stores with `derive` (🐃):
+
+```js
+const a = writable(1);
+const b = writable(2);
+const c = writable(3);
+
+const total = derive([a, b, c], ([a, b, c]) => a + b + c);
```

Comment:
This doesn't quite match the reference implementation below.
```suggestion
const total = derive([a, b, c], (a, b, c) => a + b + c);
```

Reply:
I've updated the example implementation. The reason I've steered away from 

```js
s = derive(a, b, c, (a, b, c) => a + b + c);
```

is that the current implementation allows us to do both of these, which I quite like:

```js
simple = derive([a, b], ([a, b]) => a + b);

complex = derive([a, b], async ([a, b], set) => {
  const data = await fetch_from_api(a, b);
  set(data);
});
```